### ISSUE TITLE
Migrate all protocols to use `quick-protobuf-codec`

### ISSUE THREAD

**Issue by mxinden** [libp2p/rust-libp2p#2500]
Instead of implementing Protobuf en-/decoding in each protocol and for each message exchange, one can instead implement a custom encoder abstracting the en-/decoding of Protobuf messages.
```[tasklist]
### Tasks
- [ ] https://github.com/libp2p/rust-libp2p/issues/4488
- [ ] https://github.com/libp2p/rust-libp2p/issues/4487
- [ ] https://github.com/libp2p/rust-libp2p/issues/4489
```

**Comment by nloadholtes** [comment]
Hi! I have read through this and would like to take it on. I'm still pretty new to Rust, so apologies in advance if I ask some basic questions. :smiley: 

I see that the `Encoder` and `Decoder` are now in prost-codec, and it looks like `protocols/dcutr`, `protocols/identity`, and `protocols/relay` are using it. I'll work on refactoring the other protocols packages to use prost-codec.

For this ticket, would you prefer 1 PR per package, or everything in 1 big PR? (My plan is to do 1 package in a draft PR and then have you/someone look at it to make sure I'm on the right path, and then go from there)

**Comment by thomaseizinger** [comment]
> Hi! I have read through this and would like to take it on. I'm still pretty new to Rust, so apologies in advance if I ask some basic questions. :smiley: 
> 
> I see that the `Encoder` and `Decoder` are now in prost-codec, and it looks like `protocols/dcutr`, `protocols/identity`, and `protocols/relay` are using it. I'll work on refactoring the other protocols packages to use prost-codec.
> 
> For this ticket, would you prefer 1 PR per package, or everything in 1 big PR? (My plan is to do 1 package in a draft PR and then have you/someone look at it to make sure I'm on the right path, and then go from there)

1 PR per crate would be preferable!

**Comment by mxinden** [comment]
Thanks @nloadholtes for volunteering.

**Comment by nloadholtes** [comment]
Hi @mxinden and @thomaseizinger  , I've made a WIP PR that I'd like some feedback on: #3070

Basically I'm looking at `protocols/dcutr` and trying to make sure that I can make `protocols/gossipsub` "look" like that. I ran the changes locally and the tests passed so I'm hopeful that is a good sign. :smile: 

Thank you in advance for any feedback, I'm still learning the ways of rust, but I'm really liking what I'm seeing so far.
